packages/common/src/utils/stream-utils.ts

```diff
@@ -310,6 +310,26 @@ export class StreamUtils {
     if (commitData?.capability?.p?.exp) {
       logEntry.expirationTime = Math.floor(Date.parse(commitData.capability.p.exp) / 1000)
     }
+    if (commitData.timestamp) {
+      logEntry.timestamp = commitData.timestamp
+    }
     return logEntry
   }
+
+  /**
+   * Takes a StreamState and validates that none of the commits in its log are based on expired CACAOs.
+   */
+  static checkForCacaoExpiration(state: StreamState): void {
+    const now = Math.floor(Date.now() / 1000) // convert millis to seconds
+    for (const logEntry of state.log) {
+      const timestamp = logEntry.timestamp ?? now
+      if (logEntry.expirationTime && logEntry.expirationTime < timestamp) {
+        throw new Error(
+          `CACAO expired: Commit ${logEntry.cid.toString()} of Stream ${StreamUtils.streamIdFromState(
+            state
+          ).toString()} has a CACAO that expired at ${logEntry.expirationTime}`
+        )
+      }
+    }
```

Comment:
Why not break after the first anchor is detected?

Reply:
That will only work for now since the CAS is loading the streams and validating the commits before anchoring.  Once CAS w/o Ceramic node is complete, a commit that has already expired could still be anchored

Reply:
ok